// website-contacts-extractor/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiException = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const utils_1 = require("./utils");
class ApiException extends Error {
    constructor(message, response) {
        super(message);
        Object.defineProperty(this, "response", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = "ApiException";
        this.response = response;
    }
}
exports.ApiException = ApiException;
function _createFilename(path) {
    return "output/responses/" + path + ".json";
}
function raiseIfBadException(response) {
    if (400 <= response.status && response.status < 500) {
        const data = response.data;
        const message = data === null || data === void 0 ? void 0 : data.message;
        if (message) {
            throw new ApiException(message, response);
        }
    }
}
class Api {
    constructor({ apiKey, createResponseFiles = true }) {
        Object.defineProperty(this, "_apiKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_createResponseFiles", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "API_URL", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "https://website-contacts-extractor.omkar.cloud/js/contacts"
        });
        /**
         * Initializes the API client with an API key and option to create response files.
         *
         * @param apiKey The API key for authentication.
         * @param createResponseFiles Indicates if the client should create response files for each API call. Defaults to true.
         */
        this._apiKey = apiKey;
        this._createResponseFiles = createResponseFiles;
    }
    _writeJson(filename, data) {
        /**
         * Writes data to a JSON file specified by the filename. This method only runs if the createResponseFiles flag is True.
         *
         * @param filename The filename for the JSON file to be created.
         * @param data The data to be written to the file.
         */
        if (this._createResponseFiles) {
            const path = _createFilename(filename);
            (0, utils_1.writeJsonResponse)(path, data);
            console.log(`View ${filename} response at: ./${path}`);
        }
    }
    async getContacts(website) {
        /**
         * Fetches contact information from a website.
         *
         * @param website The website URL to scrape contacts from
         * @return Contact information including emails and social media profiles
         */
        try {
            const response = await axios_1.default.get(this.API_URL, {
                params: {
                    website: website
                },
                headers: {
                    'API-Key': this._apiKey
                }
            });
            this._writeJson("get_contacts", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error) && error.response) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
}
exports.default = Api;
